### ISSUE TITLE
Question about training from scratch

### ISSUE THREAD

**Issue by sooahcho25** [halbielee/SEC_pytorch#1]
First of all, Thank you for your hard work. I am familiar with Pytorch, so your work is very helpful.
I am curious about the condition that you got the performance (mIoU 50.6049 and mIoU 49.6978) you mentioned.  When I download the weight and evaluate it, I get the same result as the performance you mentioned.

meanIOU     : 50.6049
-----------------------------------------
background  : 82.1013
aeroplane   : 59.3787
bicycle     : 25.7055
bird        : 61.6386
boat        : 26.7922
bottle      : 40.3784
bus         : 66.6121
car         : 63.3243
cat         : 75.6848
chair       : 22.2849
cow         : 54.4826
diningtable : 28.8335
dog         : 66.1762
horse       : 57.7841
motorbike   : 63.0711
person      : 53.3926
pottedplant : 32.1166
sheep       : 61.7726
sofa        : 32.0557
train       : 44.8119
tvmonitor   : 44.3050

However, if I try to train from scratch, I cannot get the same result. This is the result of testing by saving the checkpoint when the loss is minimum without changing the hyperparameter setting.

meanIOU     : 10.8535
-----------------------------------------
background  : 74.5545
aeroplane   : 12.2275
bicycle     : 0.0005
bird        : 0.0000
boat        : 1.5335
bottle      : 0.1878
bus         : 0.6897
car         : 0.0151
cat         : 8.4732

The learning loss is shown below. Even if I change the hyperparameters, the sum of losses converges to 6.xx (seed loss 2.xx, expand loss 4.xx, constrain loss 0.5xx). It seems that training is not working in my case.

![image](https://user-images.githubusercontent.com/62269746/76811474-f10d7300-6834-11ea-9ea5-ef33250575df.png)

What I'm curious about is your learning loss decreasing well? What is the saving checkpoint criteria? (minimal loss or something?)

**Comment by sooahcho25** [comment]
안녕하세요. 혹시 한국분이실 수도 있을 것 같아서, 한국어로도 문의 남깁니다. (향후 영어로 답글 달아주시면 본 글은 지우도록 하겠습니다.)

SEC를 구현하는 연습 중이었는데 올려주신 code 덕분에 많은 도움을 받았습니다.
현재 어려움을 겪고 있는 부분은 from scratch로 학습하는 경우입니다.
training loss가 6 수준으로 수렴하고, minimum loss 일때의 checkpoint를 저장하여 확인해 보았을때 아래와 같은 수준의 성능으로 나옵니다.
* seed loss 2.xx, expand loss 4.xx, constrain loss 0.5xx 로 각각 수렴합니다.

meanIOU : 10.8535
---------------------------------------
background : 74.5545
aeroplane : 12.2275
bicycle : 0.0005
bird : 0.0000
boat : 1.5335
bottle : 0.1878
bus : 0.6897
car : 0.0151
cat : 8.4732

올려주신 weight로는 mIoU 50.6049가 잘 나옵니다.
하이퍼파라미터를 변경해봐도 loss가 6으로 수렴하는데, 혹시 학습하실때 loss가 어느 정도 수준이었는지 궁금합니다.
그리고 checkpoint 저장 기준을 어떤 것으로 설정하셨는지 궁금합니다. 
저는 minimum loss일때로 저장하였는데, 다른 기준을 사용하셨을 수도 있을 것 같아서 질문 드립니다.

미리 감사드립니다.

**Comment by halbielee** [comment]
Hello, sooahcho25!
Thanks for your attention for my work. I answer your question in English for the others.

I read your question carefully, but I don't know the exact reason why it shows that low mIOU so far. :(

To begin with you question, 
1. About the values of loss
I am attaching the log for training. **All values show 1 or less.**
![SEC_loss](https://user-images.githubusercontent.com/31704942/76833599-f7204580-686e-11ea-9939-f7b8a24b359e.png)

2. About the checkpoint
I used the last checkpoint for the performance, In the [original paper](https://arxiv.org/abs/1603.06098) or [code](https://github.com/kolesman/SEC), they always use **the last checkpoint**. So, I used it as it is.

In my opinion, that so low mIOU might not be a minor hyperparamter issue.
- This could be related to the **pretrained model**. Do you use the ImageNet pretrained model? 
- If you use, which model did you use? or how did you made it?
- If not, please note that SEC uses ImageNet pretrained model.

Lastly, I did not upload the converted pretrained model for PyTorch. You can convert it from Caffe pretrained model!

If you have any more questions, please feel free to ask me.

**Comment by sooahcho25** [comment]
Thanks for the quick answer. Your answer is very helpful!
I change the pretrained model according to your advice. I didn't implement it myself, but I used what others have implemented. It differs for each pretrained model version, but the best performance I got is:

meanIOU     : 49.0975
------------------------------------------------------------------
background  : 81.3378
aeroplane   : 54.3209
bicycle     : 24.8042
bird        : 55.5517
boat        : 23.9628
bottle      : 42.5303
bus         : 68.7312
car         : 58.3029
cat         : 72.6612
chair       : 21.1893
cow         : 50.9726
diningtable : 28.1235
dog         : 65.0175
horse       : 53.1135
motorbike   : 61.8385
person      : 54.6882
pottedplant : 34.9582
sheep       : 59.8360
sofa        : 32.9742
train       : 44.0934
tvmonitor   : 42.0398

The loss graph is shown below and it seems to be learning well now. Thanks for the advice. :)

![image](https://user-images.githubusercontent.com/62269746/77142233-ca646c00-6ac2-11ea-913c-494899450405.png)

And there is one more question.
What is the difference between the two performances you mentioned? (mIoU 50.6049 vs mIoU 49.6978)

Lastly, the following is not a question, it is just my opinion.
In your code, it seems to open the image file with cv2 on training and pylab on testing. If I open with cv2 when testing, the performance of checkpoint version 1 increases slightly to mIoU 50.6073. In my past experience, there are some differences in the pixel values when the image is opened with cv2 and when opened with pylab. If you use the same image utility for train and test, it may help performance. I am checking whether it is better to learn by opening with pylab and learning by opening with cv2.

**Comment by halbielee** [comment]
@sooahcho25 ,

Thanks for the quick feed back and I am appreciated with you suggestion, as well.

**Performance** the two performances do not have any differences. I just trained the code and got two results from it :)

**pylab vs cv2** This is I what I missed. I know about the difference of value between pylab and cv2. I did not mind changing pylab to cv2. Thanks for the suggestion :)

I am glad to meet someone who research in similar area. :) It would be nice if we could talk comfortably. Thanks !

**Comment by vwonx** [comment]
@halbielee 

Hello, I used the pretrained model you provided [
vgg16_20M.pth](https://drive.google.com/a/yonsei.ac.kr/file/d/1Fs25jmy9uZJxLlWFfdOESOs5EXnAr0OK/view?usp=sharing).

But I can only get the following results, 

meanIOU     : 44.3545
=====================
background  : 80.4437
aeroplane   : 52.3944
bicycle     : 19.8133
bird        : 53.9390
boat        : 26.2848
bottle      : 35.0750
bus         : 60.6747
car         : 57.3808
cat         : 64.1090
chair       : 18.3647
cow         : 41.2378
diningtable : 21.2285
dog         : 56.5677
horse       : 47.7649
motorbike   : 52.9296
person      : 48.5751
pottedplant : 29.9387
sheep       : 50.5441
sofa        : 31.0658
train       : 41.9995
tvmonitor   : 41.1140

# The loss is shown below.

![{20C05249-50C9-2F90-D7D3-844EAF8F3F25}](https://user-images.githubusercontent.com/40535851/77752937-7a0d8100-7063-11ea-8b30-dcfbe7e1c141.jpg)

Is this problem still caused by the pretrained model? Do you change the hyperparamter in the code? Are you using the same pretrained model?

Finally, thanks for your hard work.

**Comment by sooahcho25** [comment]
Thank you for your kind reply. Thanks to you, now I can utilize SEC using pytorch. 
I am also glad to meet you. :D
Now everything works fine, so I will close this issue. 
Thank you!
